### PR TITLE
Import/export refactoring

### PR DESCRIPTION
**Pull request by vsverchinsky**
Resolves: #3863

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Comment by Paul-Licameli** [comment]
Excellent!  But still a draft.  What is missing?

**Comment by vsverchinsky** [comment]
> Excellent! But still a draft. What is missing?

Export parameters should replace global prefs and break dependency on ShuttleGUI, direct invocations of message dialogs still happening during export process, import and export process is still happening in the main thread ... I'll update parent ticket so it would be easier to understand the progress.

**Comment by Paul-Licameli** [comment]
Failing builds

File: src/export/Export.cpp [hunk 1]
```diff
@@ -90,114 +90,15 @@ bool ExportPlugin::CheckFileName(wxFileName & WXUNUSED(filename), int WXUNUSED(f
...
-   bool isext = false;
-   for (int i = index; i < GetFormatCount(); i = GetFormatCount())
+   if(index >= 0 && index < GetFormatCount())
    {
-      const auto &defext = GetExtension(i);
-      const auto &defexts = GetExtensions(i);
-      int indofext = defexts.Index(ext, false);
-      if (defext.empty() || (indofext != wxNOT_FOUND))
-         isext = true;
+      auto formatInfo = GetFormatInfo(index);
```
Comment (Paul-Licameli) [inline src/export/Export.cpp]: auto &  ?

File: src/export/Export.cpp [hunk 2]
```diff
@@ -90,114 +90,15 @@ bool ExportPlugin::CheckFileName(wxFileName & WXUNUSED(filename), int WXUNUSED(f
...
-{
-   return mFormatInfos[index].mMaxChannels;
-}
-
-bool ExportPlugin::GetCanMetaData(int index)
-{
-   return mFormatInfos[index].mCanMetaData;
-}
-
 bool ExportPlugin::IsExtension(const FileExtension & ext, int index)
```
Comment (Paul-Licameli) [inline src/export/Export.cpp]: `const` function?

File: src/export/Export.cpp [hunk 3]
```diff
@@ -418,9 +319,11 @@ bool Exporter::Process(bool selectedOnly, double t0, double t1)
    if (!CheckMix()) {
       return false;
    }
+   
+   auto exportFormatInfo = mPlugins[mFormat]->GetFormatInfo(mSubFormat);
```
Comment (Paul-Licameli) [inline src/export/Export.cpp:144]: Again, use `&`?

File: src/export/Export.cpp [hunk 4]
```diff
@@ -478,7 +381,8 @@ bool Exporter::Process(
       ++i;
       for (int j = 0; j < pPlugin->GetFormatCount(); j++)
       {
-         if (pPlugin->GetFormat(j).IsSameAs(type, false))
+         auto formatInfo = pPlugin->GetFormatInfo(j);
```
Comment (Paul-Licameli) [inline src/export/Export.cpp:198]: again

File: src/export/Export.cpp [hunk 5]
```diff
@@ -603,9 +507,9 @@ bool Exporter::GetFilename()
          ++i;
          for (int j = 0; j < pPlugin->GetFormatCount(); j++)
          {
-            auto mask = pPlugin->GetMask(j);
-            fileTypes.insert( fileTypes.end(), mask.begin(), mask.end() );
-            if (mPlugins[i]->GetFormat(j) == defaultFormat) {
+            auto formatInfo = pPlugin->GetFormatInfo(j);
```
Comment (Paul-Licameli) [inline src/export/Export.cpp]: again

File: src/export/Export.cpp [hunk 6]
```diff
@@ -603,9 +507,9 @@ bool Exporter::GetFilename()
...
+            fileTypes.insert(fileTypes.end(), { formatInfo.mDescription, formatInfo.mExtensions });
```
Comment (Paul-Licameli) [inline src/export/Export.cpp]: Correctness of this change assumes `SetMask` was nowhere used -- and that is true

File: src/export/ExportCL.cpp [hunk 7]
```diff
@@ -343,15 +346,18 @@ class ExportCL final : public ExportPlugin
...
-   SetMaxChannels(255,0);
-   SetCanMetaData(false,0);
-   SetDescription(XO("(external program)"),0);
+   return 1;
+}
+
+FormatInfo ExportCL::GetFormatInfo(int) const
+{
+   return {
+      wxT("CL"), XO("(external program)"), {""}, {}, 255, false
```
Comment (Paul-Licameli) [inline src/export/ExportCL.cpp]: The `{}` here and elsewhere initializes `mMask` but that member is nowhere used, so simplify

File: src/export/ExportPCM.cpp [hunk 8]
```diff
@@ -409,29 +408,44 @@ class ExportPCM final : public ExportPlugin
...
-   SetDescription(XO("Other uncompressed files"), selformat);
-   SetCanMetaData(true, selformat);
-   SetMaxChannels(255, selformat);
+      return {
+         sf_header_shortname(si.format),
+         XO("Other uncompressed files"),
+         { sf_header_extension(si.format) },
+         {},
+         static_cast<unsigned>(si.channels),
+         true
```
Comment (Paul-Licameli) [inline src/export/ExportPCM.cpp:440]: Not the literals "LIBSNDFILE" and `255` as before.  Is this right?

Comment (Paul-Licameli) [reply src/export/ExportPCM.cpp:440]: Commit comment should mention that you are correcting this case, not just refactoring

File: src/export/ExportPCM.cpp [hunk 9]
```diff
@@ -409,29 +408,44 @@ class ExportPCM final : public ExportPlugin
...
-   }
+      SF_INFO si = {};
+
+      si.format = LoadOtherFormat() & SF_FORMAT_TYPEMASK;
+      si.format |= LoadEncoding(si.format);
+
+      for (si.channels = 1; sf_format_check(&si); si.channels++)
+      {
+         // just counting
+      }
```
Comment (Paul-Licameli) [inline src/export/ExportPCM.cpp:433]: Did you neglect `--si.channels;` after the loop?

File: src/export/ExportCL.cpp (hunk 7 above)
Comment (Paul-Licameli) [reply src/export/ExportCL.cpp]: ... And so you did in the second commit.  Just squash it.

File: src/export/Export.cpp [hunk 10]
```diff
@@ -66,13 +47,13 @@
...
 #include "FileNames.h"
 #include "widgets/HelpSystem.h"
 #include "widgets/ProgressDialog.h"
 #include "wxFileNameWrapper.h"
 
+#include "ExportMixerDialog.h"
+
 //----------------------------------------------------------------------------
 // ExportPlugin
 //----------------------------------------------------------------------------
```
Comment (Paul-Licameli) [inline src/export/Export.cpp]: Move (or just delete) the banner comments too, for the moved classes

Comment (Paul-Licameli) [reply src/export/Export.cpp]: The commit comment can also mention ExportPlugin

File: src/export/ExportMixerPanel.h [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+///\class ExportMixerPanel
+///\brief Panel that displays mixing for advanced mixing option.
+class ExportMixerPanel final : public wxPanelWrapper
+{
+public:
+   ExportMixerPanel( wxWindow *parent, wxWindowID id,
+         MixerOptions::Downmix *mixerSpec, wxArrayString trackNames,
+         const wxPoint& pos = wxDefaultPosition,
+         const wxSize& size = wxDefaultSize);
```
Comment (Paul-Licameli) [inline src/export/ExportMixerPanel.h:31]: Not obvious in this large commit -- this is not just copied and pasted, but also the type of the third argument changes.  Please mention that in the commit comment.

Comment (Paul-Licameli) [reply src/export/ExportMixerPanel.h:31]: Oh but I see now -- same type, just another alias.  So never mind.

File: src/import/ImportProgressListener.h [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+   enum class ImportResult {
+      Success,
+      Error,
+      Cancelled
+   };
+   
+   virtual ~ImportProgressListener() = default;
+   
+   ///Called by Importer when it attempts to import file using registered ImportPlugin instance
+   ///Could be called more than once, but for each call there will be a complimentary call to OnImportResult
```
Comment (Paul-Licameli) [inline src/import/ImportProgressListener.h:30]: should be "complementary"

File: src/import/ImportProgressListener.h [hunk 13]
```diff
@@ -0,0 +1,41 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  ImportProgressListener.h
+
+  V. Sverchinsky
+
+*//****************************************************************/
```
Comment (Paul-Licameli) [inline src/import/ImportProgressListener.h:9]: As a matter of form I'd prefer that there is also a .cpp including this so we are sure the header has all its compilation dependencies, and the destructor should have a non-inline definition there.

File: src/ProjectFileManager.cpp [hunk 14]
```diff
@@ -1204,6 +1207,55 @@ bool ImportProject(AudacityProject &dest, const FilePath &fileName)
...
+         importFileHandle.SetStreamUsage(0,TRUE);
+      return true;
+   }
+   
+   void OnImportProgress(double progress) override
+   {
+      constexpr double ProgressSteps { 1000.0 };
+      if(!mProgressDialog)
+      {
+         wxFileName ff( mImportFileHandle->GetFilename() );
```
Comment (Paul-Licameli) [inline src/ProjectFileManager.cpp:1238]: Should I worry about the safety of that `->`? Should there be a shared or weak pointer?  Or should there be comments in the base class header explaining the assumptions that derived classes may make about lifetimes of `ImportFileHandle`?

File: src/import/ImportWavPack.cpp [hunk 15]
```diff
@@ -228,20 +233,26 @@ ProgressResult WavPackImportFileHandle::Import(WaveTrackFactory *trackFactory, T
...
    if (WavpackGetNumErrors(mWavPackContext))
       AudacityMessageBox( XO( "Encountered %d errors decoding WavPack file!" ).Format( WavpackGetNumErrors(mWavPackContext) ),
                           XO( "WavPack Importer" ), wxOK | wxICON_EXCLAMATION | wxCENTRE);
 
-   if (updateResult != ProgressResult::Stopped && updateResult != ProgressResult::Cancelled
-         && totalSamplesRead < mNumSamples)
-      updateResult = ProgressResult::Failed;
-
-   if (updateResult == ProgressResult::Failed || updateResult == ProgressResult::Cancelled)
-      return updateResult;
```
Comment (Paul-Licameli) [inline src/import/ImportWavPack.cpp:243]: In the old logic, there was a distinction between stopping and cancelling.  In case of stopping, there was no error reported, even if `totalSamplesRead < mNumSamples`.

In the new logic, this changes.

Comment (Paul-Licameli) [reply src/import/ImportWavPack.cpp:243]: Progress dialogs had a Stop button distinct from Cancel in the previous version, but with this commit they don't.

Is this change intended?

File: src/import/ImportFFmpeg.cpp [hunk 16]
```diff
@@ -522,12 +522,11 @@ ProgressResult FFmpegImportFileHandle::Import(WaveTrackFactory *trackFactory,
    }
    // This is the heart of the importing process
    // The result of Import() to be returned. It will be something other than zero if user canceled or some error appears.
-   auto res = ProgressResult::Success;
-
+   auto hasError {false};
```
Comment (Paul-Licameli) [inline src/import/ImportFFmpeg.cpp:525]: This is an unusual way to write it.

Funny C++ fact:  if you change it to
`auto hasError = {false};`

then instead they type is deduced as `std::initializer_list<bool>` and compilation fails because that is not convertible too bool.

File: src/import/ImportOGG.cpp [hunk 17]
```diff
@@ -336,19 +340,22 @@ ProgressResult OggImportFileHandle::Import(
 
          samplesSinceLastCallback += samplesRead;
          if (samplesSinceLastCallback > SAMPLES_PER_CALLBACK) {
-            updateResult = mProgress->Update(ov_time_tell(mVorbisFile.get()),
-               ov_time_total(mVorbisFile.get(), bitstream));
+            progressListener.OnImportProgress(ov_time_tell(mVorbisFile.get()) / ov_time_total(mVorbisFile.get(), bitstream));
```
Comment (Paul-Licameli) [inline src/import/ImportOGG.cpp:343]: another denominator to check for 0

File: src/export/ExportFileDialog.cpp [hunk 18]
```diff
@@ -0,0 +1,345 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  ExportFileDialog.cpp
+
+  Dominic Mazzon
```
Comment (Paul-Licameli) [inline src/export/ExportFileDialog.cpp:7]: Mazzoni

File: src/export/ExportPCM.cpp [hunk 19]
```diff
@@ -33,6 +33,7 @@
 #include "../widgets/ProgressDialog.h"
 #include "../widgets/wxWidgetsWindowPlacement.h"
 #include "wxFileNameWrapper.h"
+#include "ExportFileDialog.h"
```
Comment (Paul-Licameli) [inline src/export/ExportPCM.cpp:36]: I thought this was unnecessary for the build, but now I see it's needed for `AUDACITY_FILE_SUFFIX_EVENT`

That is one place where I omitted to change wxEvents to Observer messages, but then this is a case of wxWidgets dialogs communicating to each other.

File: src/import/ImportPCM.cpp [hunk 20]
```diff
@@ -373,20 +385,16 @@ ProgressResult PCMImportFileHandle::Import(WaveTrackFactory *trackFactory,
...
 
-         updateResult = mProgress->Update(
-            framescompleted.as_long_long(),
-            fileTotalFrames.as_long_long()
-         );
-         if (updateResult != ProgressResult::Success)
-            break;
-
-      } while (block > 0);
+         progressListener.OnImportProgress(framescompleted.as_double() / fileTotalFrames.as_double());
```
Comment (Paul-Licameli) [inline src/import/ImportPCM.cpp:388]: And one more denominator.

Maybe `OnImportProgress` should take numerator and denominator as two arguments, and you can write the check for zero division in one place only.

File: src/export/ExportFileDialog.cpp [hunk 21]
```diff
@@ -0,0 +1,345 @@
...
+                             const wxString& defaultName,
+                             const wxString& defaultFormatName,
+                             const TranslatableString& title,
+                             const wxPoint& pos,
+                             const wxSize& sz,
+                             const wxString& name)
+{
+   wxString formatName = defaultFormatName;
+   if(defaultFormatName.empty())
+      formatName = gPrefs->Read(wxT("/Export/Format"),
```
Comment (Paul-Licameli) [inline src/export/ExportFileDialog.cpp:119]: It would be good to make a StringSetting now and not repeat the literal path string

File: src/export/Export.cpp [hunk 22]
```diff
@@ -800,7 +471,6 @@ bool Exporter::ProcessFromTimerRecording(bool selectedOnly,
    mFilename = fnFile;
    mFormat = iFormat;
    mSubFormat = iSubFormat;
-   mFilterIndex = iFilterIndex;
```
Comment (Paul-Licameli) [inline src/export/Export.cpp:1042]: With this deletion, iFilterIndex becomes unused.

I expected then that Timer record would achieve this side effect otherwise, by a call to configure the exporter, but I don't see it.

Comment (Paul-Licameli) [reply src/export/Export.cpp:1042]: If that is correct, then `m_iAutoExportFilterIndex` is a useless member of `TimerRecordDialog`.

Comment (Paul-Licameli) [reply src/export/Export.cpp:1042]: And if that is removed, then `Exporter::GetAutoExportFilterIndex()` will be called nowhere.

File: src/export/ExportFileDialog.cpp [hunk 23]
```diff
@@ -0,0 +1,345 @@
...
+                                         wxT("WAV"));
+   
+   wxFileName filename;
+   //Bug 1304: Set a default path if none was given.  For Export.
+   filename.SetPath(FileNames::FindDefaultPath(FileNames::Operation::Export));
+   filename.SetFullName(defaultName);
```
Comment (Paul-Licameli) [inline src/export/ExportFileDialog.cpp:125]: I think this has `SetFullName` where the old version had `SetName` in `Exporter::GetFileName()`.  Why the change?

File: src/export/ExportFileDialog.cpp [hunk 24]
```diff
@@ -0,0 +1,345 @@
...
+   else
+      filename.SetFullName(defaultName);
+   
+   if (!filename.HasExt())
+   {
+      auto defext = plugins[pluginIndex]->GetFormatInfo(formatIndex).mExtensions[0].Lower();
+      filename.SetExt(defext);
+   }
+   
+   wxTabTraversalWrapper<FileDialog>::Create(parent,
```
Comment (Paul-Licameli) [inline src/export/ExportFileDialog.cpp:96]: Why not `FileDialogWrapper` as in the original?

File: src/export/ExportFileDialog.cpp [hunk 25]
```diff
@@ -0,0 +1,345 @@
...
+                                             filename.GetFullName(),
+                                             FileNames::FormatWildcard(fileTypes),
+                                             wxRESIZE_BORDER | wxFD_SAVE | wxFD_OVERWRITE_PROMPT,
+                                             pos, sz, name);
+   
+   Bind(wxEVT_FILECTRL_FILTERCHANGED, &ExportFileDialog::OnFilterChanged, this);
+   SetUserPaneCreator(CreateUserPaneCallback, (wxUIntPtr) this);
+}
+
+int ExportFileDialog::Export(wxWindow* parent,
```
Comment (Paul-Licameli) [inline src/export/ExportFileDialog.cpp:108]: `Export` is a misleading name.  This function only chooses a path before export really happens.

File: src/export/ExportFileDialog.cpp [hunk 26]
```diff
@@ -0,0 +1,345 @@
...
+            auto prompt =
+               XO("Are you sure you want to export the file as \"%s\"?\n")
+                  .Format( filename.GetFullName() );
+
+            int action = AudacityMessageBox(
+               prompt,
+               XO("Warning"),
+               wxYES_NO | wxICON_EXCLAMATION);
+            if (action != wxYES) {
+               return false;
```
Comment (Paul-Licameli) [inline src/export/ExportFileDialog.cpp:196]: I think this should be `continue` to conserve the old logic

File: src/export/ExportFileDialog.cpp [hunk 27]
```diff
@@ -0,0 +1,345 @@
...
+            }
+         }
+
+         filename.SetExt(defext);
+      }
+      
+      if (!plugins[pluginIndex]->CheckFileName(filename, formatIndex))
+      {
+         return false;
```
Comment (Paul-Licameli) [inline src/export/ExportFileDialog.cpp:205]: `continue` here too

File: src/export/ExportFileDialog.cpp [hunk 28]
```diff
@@ -0,0 +1,345 @@
...
+            wxYES_NO | wxICON_EXCLAMATION);
+         if (action != wxYES) {
+            return false;
+         }
+      }
+
+      if (filename.GetFullPath().length() >= 256) {
+         AudacityMessageBox(
+            XO( "Sorry, pathnames longer than 256 characters not supported.") );
+         return false;
```
Comment (Paul-Licameli) [inline src/export/ExportFileDialog.cpp:225]: `continue` here too

**Comment by Paul-Licameli** [comment]
This round of review is complete!

File: src/export/ExportUtils.h [hunk 29]
```diff
@@ -36,5 +39,6 @@ class ExportUtils final
    static void InitProgress(std::unique_ptr<BasicUI::ProgressDialog> &pDialog,
          const wxFileNameWrapper &title, const TranslatableString &message);
 
+   static TrackIterRange<const WaveTrack> FindExportWaveTracks(const TrackList& tracks, bool selectedOnly);
```
Comment (Paul-Licameli) [inline src/export/ExportUtils.h:42]: Do you understand the difference between `SampleTrack` and `WaveTrack`?  The first is an abstract base class of the second.  (And `WritableSampleTrack` is between them.)

I think these utilities would be better defined in terms of the more general interface.  We don't yet have other concrete subclasses of `SampleTrack` but it may be useful later.

`SampleTrack` does not expose the notion of a clip within the track, but I don't see that you require knowledge of subdivision into clips.

File: src/export/ExportUtils.cpp [hunk 30]
```diff
@@ -59,3 +54,13 @@ void ExportUtils::InitProgress(std::unique_ptr<BasicUI::ProgressDialog> &pDialog
    return InitProgress(
       pDialog, Verbatim( title.GetName() ), message );
 }
+
+//TODO: used in many places in anticipation that Exporter yields same result, fix that
```
Comment (Paul-Licameli) [inline src/export/ExportUtils.cpp:58]: I don't understand the comment.  What "same result?"

File: src/export/ExportUtils.cpp [hunk 31]
```diff
@@ -59,3 +54,13 @@ void ExportUtils::InitProgress(std::unique_ptr<BasicUI::ProgressDialog> &pDialog
...
+TrackIterRange<const WaveTrack> ExportUtils::FindExportWaveTracks(const TrackList& tracks, bool selectedOnly)
+{
+   bool anySolo = !(( tracks.Any<const WaveTrack>() + &WaveTrack::GetSolo ).empty());
+
+   return tracks.Any<const WaveTrack>()
+            + ( selectedOnly ? &Track::IsSelected : &Track::Any  )
+            - ( anySolo ? &WaveTrack::GetNotSolo :  &WaveTrack::GetMute);
```
Comment (Paul-Licameli) [inline src/export/ExportUtils.cpp:65]: `Track` which is the base of `SampleTrack` does expose selectedness.

But it is `PlayableTrack` that defines mute and solo and it is not a base of `SampleTrack`.

Yet you still might apply `Filter` to the track range to get a range of `SampleTrack`.

Maybe `PlayableTrack` should instead become a mix-on property of tracks that you discover with a `dynamic_cast`?  That's a bigger idea I should think about for future.

Comment (Paul-Licameli) [reply src/export/ExportUtils.cpp:65]: Feel free to ignore all the suggestions about `SampleTrack` for a draft.  But remember the idea.

File: src/export/Export.cpp [hunk 32]
```diff
@@ -385,22 +373,26 @@ bool Exporter::CheckMix(bool prompt /*= true*/ )
...
-                           exportedChannels,
-                           NULL,
-                           1,
-                           XO("Advanced Mixing Options"));
-      if (prompt) {
-         if (md.ShowModal() != wxID_OK) {
+      auto exportTracks = ExportUtils::FindExportWaveTracks(TrackList::Get( *mProject ), mSelectedOnly);
+      mMixerSpec = std::make_unique<MixerOptions::Downmix>(exportTracks.size(),
+                                                           std::clamp(exportedChannels,
+                                                                      // STF (April 2016): AMR (narrowband) and MP3 may export 1 channel.
```
Comment (Paul-Licameli) [inline src/export/Export.cpp]: No preservation of the other old comments from JKC? (James)

**Comment by Paul-Licameli** [comment]
That's all for the first new commit

File: src/export/ExportMP3.cpp [hunk 33]
```diff
@@ -1757,15 +1758,6 @@ bool ExportMP3::CheckFileName(wxFileName & WXUNUSED(filename), int WXUNUSED(form
    return true;
 }
 
-int ExportMP3::SetNumExportChannels()
-{
-   bool mono;
-   gPrefs->Read(wxT("/FileFormats/MP3ForceMono"), &mono, 0);
```
Comment (Paul-Licameli) [inline src/export/ExportMP3.cpp:1755]: If you change lines like this, also take the opportunity to eliminate one more repeated string literal path with a `BoolSetting` object.

File: src/TimerRecordDialog.cpp [hunk 34]
```diff
@@ -591,9 +591,21 @@ int TimerRecordDialog::ExecutePostRecordActions(bool bWasStopped) {
 
    // Do Automatic Export?
    if (m_bAutoExportEnabled) {
+      const auto& tracks = TrackList::Get(mProject);
       Exporter e{ mProject };
+      
+      bool skipSilenceAtBeginning;
+      gPrefs->Read(wxT("/AudioFiles/SkipSilenceAtBeginning"),
```
Comment (Paul-Licameli) [inline src/TimerRecordDialog.cpp:611]: Another opportunity to use `BoolSetting`.

But be careful:  this one is used as an argument to `TieCheckBox`.

When that happens, you also need to use `Invalidate` in the `Commit()` member function of the preference panel.

I intend to remove that need for `Invalidate`, but it requires first using a named Setting object in every call of the `Tie...` functions, in Preferences, and also in Export dialogs, and also in SoundActivatedRecord.cpp

File: src/menus/FileMenus.cpp [hunk 35]
```diff
@@ -34,6 +34,35 @@
...
+bool ExportWithPrompt(AudacityProject &project,
+                      Exporter &exporter,
+                      const FileExtension& format,
+                      double t0, double t1, bool selectedOnly)
+{
+   bool skipSilenceAtBeginning;
+   gPrefs->Read(wxT("/AudioFiles/SkipSilenceAtBeginning"),
+                                   &skipSilenceAtBeginning, false);
+   
+   if(!exporter.SetExportRange(t0, t1, false, skipSilenceAtBeginning))
```
Comment (Paul-Licameli) [inline src/menus/FileMenus.cpp:52]: Did you mean to pass `selectedOnly` for the third argument?

File: src/TimerRecordDialog.cpp [hunk 36]
```diff
@@ -591,9 +591,21 @@ int TimerRecordDialog::ExecutePostRecordActions(bool bWasStopped) {
...
+                                      &skipSilenceAtBeginning, false);
+      
+      if(!e.SetExportRange(0, tracks.GetEndTime(), false, skipSilenceAtBeginning))
+      {
+         //But there should be at least one recorded track...
+         ShowExportErrorDialog(
+            ":576", XO("All audio is muted."), XO("Warning"), false);
```
Comment (Paul-Licameli) [inline src/TimerRecordDialog.cpp:618]: You lifted this UI out of the testing logic,but now it duplicates a string literal in two places.  Is there a good file in which to put a single occurrence of the literal?

File: src/menus/FileMenus.cpp [hunk 37]
```diff
@@ -34,6 +34,35 @@
 // private helper classes and functions
 namespace {
 
+bool ExportWithPrompt(AudacityProject &project,
+                      Exporter &exporter,
+                      const FileExtension& format,
+                      double t0, double t1, bool selectedOnly)
+{
```
Comment (Paul-Licameli) [inline src/menus/FileMenus.cpp:47]: This new function shows the file dialog after the setting of export range, but previous to this commit, the two things were in the opposite order (as they remain for timer recording).

So there may be an unintended, though unimportant, change in behavior.  Such as giving the warning dialog before any file selection.

Better to change no behavior unless really meaning to.

File: src/export/ExportPCM.cpp (hunk 9 above)
Comment (vsverchinsky) [reply src/export/ExportPCM.cpp:431]: Thanks!

File: src/ProjectFileManager.cpp (hunk 14 above)
Comment (vsverchinsky) [reply src/ProjectFileManager.cpp:1245]: Aren't comments for `ImportProgressListener` enough?

File: src/import/ImportWavPack.cpp (hunk 15 above)
Comment (vsverchinsky) [reply src/import/ImportWavPack.cpp:243]: I think that the major distinction between them is that the data that has been been read by the time Stop button was pressed is preserved, but mpg123 importer seem to be an exception. Will update this commit so that this behavior is preserved.

File: src/import/ImportAUP.cpp [hunk 38]
```diff
@@ -525,14 +526,19 @@ bool AUPImportFileHandle::Open()
    return false;
 }
 
+void AUPImportFileHandle::Cancel()
+{
```
[the remark this answers is not in the document]
Comment (vsverchinsky) [reply src/import/ImportAUP.cpp:520]: Moved them to `ImportPluginHandleEx`, though I don't want to convert it's type to atomic in advance

File: src/ProjectFileManager.cpp [hunk 39]
```diff
@@ -1204,6 +1207,55 @@ bool ImportProject(AudacityProject &dest, const FilePath &fileName)
...
+         auto title = XO("Importing %s").Format(  mImportFileHandle->GetFileDescription() );
+         mProgressDialog = BasicUI::MakeProgress(title, Verbatim(ff.GetFullName()), BasicUI::ProgressShowCancel);
+      }
+      auto result = mProgressDialog->Poll(progress * ProgressSteps, ProgressSteps);
```
[the remark this answers is not in the document]
Comment (vsverchinsky) [reply src/ProjectFileManager.cpp:1249]: Probably, but similarly I don't want to do it in advance

File: src/export/ExportFileDialog.cpp (hunk 23 above)
Comment (vsverchinsky) [reply src/export/ExportFileDialog.cpp:125]: This should not change the behavior, considering that currently `ExportFileDialog::Show` is always called with project name without extension, but in context of `ExportFileDialog` I think it make more sense to allow user to specify full file name.

File: src/menus/FileMenus.cpp (hunk 35 above)
Comment (vsverchinsky) [reply src/menus/FileMenus.cpp:65]: Yes, thanks!

File: src/TimerRecordDialog.cpp (hunk 36 above)
Comment (vsverchinsky) [reply src/TimerRecordDialog.cpp:619]: No, but if I understand correctly this really shouldn't happen.

File: src/menus/FileMenus.cpp (hunk 37 above)
Comment (vsverchinsky) [reply src/menus/FileMenus.cpp:60]: The behavior is restored in one of the latest commits

File: src/export/Export.cpp [hunk 40]
```diff
@@ -300,91 +289,6 @@ void Exporter::FixFilename()
...
-      if(prompt)
-      {
-         ExportMixerDialog md(exportTracks, mMixerSpec.get(),
-                              NULL,
-                              1,
-                              XO("Advanced Mixing Options"));
-         if(md.ShowModal() != wxID_OK)
-            return false;
-      }
-      mChannels = mMixerSpec->GetNumChannels();
```
Comment (Paul-Licameli) [inline src/export/Export.cpp:888]: At the commit "Move `ExportMixerDialog` invocation and mix-related warnings", this line is deleted, and I don't see where it moves.

File: src/export/Export.cpp [hunk 41]
```diff
@@ -485,6 +377,52 @@ bool Exporter::SetAutoExportOptions() {
...
+                                                                   // state the number of channels - so we assume there are always at least two.
+                                                                   // The downside is that if someone is exporting to a mono device, the dialog
+                                                                   // will allow them to output to two channels. Hmm.  We may need to revisit this.
+                                                                   // STF (April 2016): AMR (narrowband) and MP3 may export 1 channel.
+                                                                   1u,
+                                                                   MaxExportChannels));
+   return mMixerSpec.get();
+}
+
+Exporter::DownMixMode Exporter::SetUseStereoOrMonoOutput()
```
Comment (Paul-Licameli) [inline src/export/Export.cpp:387]: The return value of this function is not (yet?) used at the commit that introduces the function.

File: src/export/Export.cpp [hunk 42]
```diff
@@ -336,21 +321,6 @@ bool Exporter::ExportTracks(
    return success;
 }
 
-bool Exporter::ProcessFromTimerRecording()
```
Comment (Paul-Licameli) [inline src/export/Export.cpp]: I'm pleased to see this function unified with `Process`.  It was an ugly violation of open-closed to see this as part of the Exporter interface.

Fix the spelling in the commit comment: "became"

File: src/ProjectFileManager.cpp [hunk 43]
```diff
@@ -1211,8 +1211,15 @@ bool ImportProject(AudacityProject &dest, const FilePath &fileName)
 class ImportProgress final
    : public ImportProgressListener
 {
+   wxWeakRef<AudacityProject> mProject;
 public:
    
+   ImportProgress(AudacityProject& project)
```
Comment (Paul-Licameli) [inline src/ProjectFileManager.cpp:1217]: `explicit`

File: libraries/lib-import-export/PlainExportOptionsEditor.cpp [hunk 44]
```diff
@@ -0,0 +1,106 @@
...
+      else if(auto val = std::get_if<int>(&it->second))
+         config.Read(mConfigKeys[index], val);
+      else if(auto val = std::get_if<double>(&it->second))
+         config.Read(mConfigKeys[index], val);
+      else if(auto val = std::get_if<std::string>(&it->second))
+      {
+         wxString wxstr;
+         if(config.Read(mConfigKeys[index], &wxstr))
+            *val = wxstr.ToStdString();
+      }
```
Comment (Paul-Licameli) [inline libraries/lib-import-export/PlainExportOptionsEditor.cpp:81]: I could use my new utility like this, avoiding repetition, and requiring no rewriting in case there is ever a need
to add more variant alternatives that the generic lambda handles:

```
Visit(Variant::MakeVariant(
   [](auto &val) { config.Read(mConfigKeys[index], &val); },
   [](wxString &val) {
      wxString wxStr;
      if(config.Read(mConfigKeys[index], &wxstr))
         *val = wxstr.ToStdString();
   }
), it->second);
```

Similarly in other places where you use `get_if`

**Comment by saintmatthieu** [comment]
I just tried to open a wav file through the File>Recent menu, which yielded 
image ; Maybe I'll find out that it's intended, but I'm skeptical.

**Comment by vsverchinsky** [comment]
Closed in favor of https://github.com/audacity/audacity/pull/4668
